<img src=".\images\Logo.png" width=150 align="left" /> <img src=".\images\Logo2.jpg" width=450 align="right" />


# <center><font color= #1e8449 > <b>CUxx_xxxxx</font></center>

<font color='steelblue'><b>Citizenlab Data Science Methodology > II - Data Processing Domain </font>
***
> # <font color='steelblue'> <b>05.- Data Collection</font>

<font color='steelblue'>Data Collection is the process to obtain and generate (if required) necessary data to model the problem.</font>

### 19. Unir todos los datos en un data.frame con repeticiones

* Unión de todas las tablas donde se reptirán los datos que no tengan la máxima resolución







<h1>Table of Contents<span class="tocSkip"></span></h1>
<div class="toc"><ul class="toc-item"><li><span><a href="#Settings" data-toc-modified-id="Settings-1"><font color="green">Settings</font></a></span></li><li><span><a href="#Data-Load" data-toc-modified-id="Data-Load-2"><font color="green">Data Load</font></a></span></li><li><span><a href="#ETL-Processes" data-toc-modified-id="ETL-Processes-3">ETL Processes</a></span><ul class="toc-item"><li><span><a href="#Import-data-from:-CSV,-Excel,-Tab,-JSON,-SQL,-and-Parquet-files" data-toc-modified-id="Import-data-from:-CSV,-Excel,-Tab,-JSON,-SQL,-and-Parquet-files-3.1">Import data from: CSV, Excel, Tab, JSON, SQL, and Parquet files</a></span></li></ul></li><li><span><a href="#Synthetic-Data-Generation" data-toc-modified-id="Synthetic-Data-Generation-4">Synthetic Data Generation</a></span></li><li><span><a href="#Fake-Data-Generation" data-toc-modified-id="Fake-Data-Generation-5">Fake Data Generation</a></span></li><li><span><a href="#Open-Data" data-toc-modified-id="Open-Data-6">Open Data</a></span></li><li><span><a href="#Data-Save" data-toc-modified-id="Data-Save-7"><font color="green">Data Save</font></a></span></li><li><span><a href="#Main-Conclusions" data-toc-modified-id="Main-Conclusions-8"><font color=" #2874a6 ">Main Conclusions</font></a></span></li><li><span><a href="#Main-Actions" data-toc-modified-id="Main-Actions-9"><font color=" #2874a6 ">Main Actions</font></a></span><ul class="toc-item"><li><ul class="toc-item"><li><span><a href="#Acciones-done" data-toc-modified-id="Acciones-done-9.0.1"><font color=" #2874a6 ">Acciones done</font></a></span></li><li><span><a href="#Acctions-to-perform" data-toc-modified-id="Acctions-to-perform-9.0.2"><font color=" #2874a6 ">Acctions to perform</font></a></span></li></ul></li></ul></li></ul></div>

## <font color='green'>Settings</font>

### Packages to use


- {tcltk} para selección interactiva de archivos locales
- {readr} para leer y escribir archivos csv
- {dplyr} para explorar datos


In [1]:
library(readr)
library(dplyr)





Attaching package: 'dplyr'


The following objects are masked from 'package:stats':

    filter, lag


The following objects are masked from 'package:base':

    intersect, setdiff, setequal, union




### Paths

In [2]:
iPath <- "Data/Input/"
oPath <- "Data/Output/"

## <font color='green'>Data Load</font>

If there are more than one input file, make as many sections as files to import.

Instrucciones
- Los ficheros de entrada del proceso están siempre en Data/Input/.   
- Si hay más de un fichero de entrada, se crean tantos objetos iFile_xx y file_data_xx como ficheros de entrada (xx número correlativo con dos dígitos, rellenar con ceros a la izquierda)

1. Datos por zona y semana

<font color='tomato'><b> OPCION A:</b> Seleccionar fichero en ventana para mayor comodidad</font> 

Data load using the {tcltk} package. Ucomment the line if not using this option

In [ ]:
# file_data_01 <- tcltk::tk_choose.files(multi = FALSE)

<font color='tomato'><b> OPCION B:</b> Especificar el nombre de archivo</font>



In [3]:
iFile_01 <- "CU_04_05_16_vacunacion_gripe.csv"
file_data_01 <- paste0(iPath, iFile_01)

if(file.exists(file_data_01)){
    cat("Se leerán datos del archivo: ", file_data_01)
} else{
    warning("Cuidado: el archivo no existe.")
}


Se leer<U+00E1>n datos del archivo:  Data/Input/CU_04_05_16_vacunacion_gripe.csv

#### Data file to dataframe

Usar la función adecuada según el formato de entrada (xlsx, csv, json, ...)

In [4]:
data_01 <- read_csv(file_data_01)


Rows: 21736 Columns: 19
-- Column specification --------------------------------------------------------
Delimiter: ","
chr  (2): GEOCODIGO, DESBDT
dbl (17): ano, semana, n_vacunas, n_citas, tmed, prec, velmedia, presMax, be...

i Use `spec()` to retrieve the full column specification for this data.
i Specify the column types or set `show_col_types = FALSE` to quiet this message.


Estructura de  los datos:

In [5]:
data_01 |> glimpse()

Rows: 21,736
Columns: 19
$ GEOCODIGO <chr> "001", "001", "001", "001", "001", "001", "001", "001", "001~
$ DESBDT    <chr> "Abrantes", "Abrantes", "Abrantes", "Abrantes", "Abrantes", ~
$ ano       <dbl> 2021, 2021, 2021, 2021, 2021, 2021, 2021, 2021, 2021, 2021, ~
$ semana    <dbl> 36, 37, 38, 39, 40, 41, 42, 43, 44, 45, 46, 47, 48, 49, 50, ~
$ n_vacunas <dbl> 0, 0, 0, 0, 0, 328, 344, 353, 371, 341, 389, 349, 402, 350, ~
$ n_citas   <dbl> 0, 0, 0, 0, 0, 305, 327, 328, 341, 334, 373, 332, 370, 336, ~
$ tmed      <dbl> 23.822231, 20.160014, 18.551058, 18.815387, 17.494475, 17.56~
$ prec      <dbl> -0.0063023484, 3.5537258042, 3.9769052178, 1.4806472324, -0.~
$ velmedia  <dbl> 3.573728, 2.494744, 3.316148, 2.384262, 1.850839, 1.527085, ~
$ presMax   <dbl> 940.9841, 940.7610, 943.1540, 944.6697, 944.2973, 943.0733, ~
$ benzene   <dbl> 0.1713567, 0.1573829, 0.1858059, 0.1486437, 0.1428037, 0.141~
$ co        <dbl> 0.1680325, 0.2138607, 0.2034376, 0.2399882, 0.2693450, 0.283~
$ no        <db

Muestra de datos:

In [6]:
data_01 |> slice_head(n = 5)

GEOCODIGO DESBDT   ano  semana n_vacunas n_citas tmed     prec        
1 001       Abrantes 2021 36     0         0       23.82223 -0.006302348
2 001       Abrantes 2021 37     0         0       20.16001  3.553725804
3 001       Abrantes 2021 38     0         0       18.55106  3.976905218
4 001       Abrantes 2021 39     0         0       18.81539  1.480647232
5 001       Abrantes 2021 40     0         0       17.49447 -0.033790900
  velmedia presMax  benzene   co        no        no2      nox      o3      
1 3.573728 940.9841 0.1713567 0.1680325  4.098371 20.09480 26.48135 50.03434
2 2.494744 940.7610 0.1573829 0.2138607  6.515572 27.42594 37.45944 42.41281
3 3.316148 943.1540 0.1858059 0.2034376  5.477654 20.74836 25.61128 56.29918
4 2.384262 944.6697 0.1486437 0.2399882  9.593391 37.08524 52.43745 46.79483
5 1.850839 944.2973 0.1428037 0.2693450 18.860535 40.19475 74.04903 41.06600
  pm10     pm2.5     so2     
1 17.44765  3.008675 6.861545
2 17.65840 10.083070 6.589638
3 12.84444  7.218588 4.364304
4 16.39590  9.426029 3.123598
5 14.90938  8.131753 1.291137

<font color='tomato'><b> OPCION A:</b> Seleccionar fichero en ventana para mayor comodidad</font> 

Data load using the {tcltk} package. Ucomment the line if not using this option

In [ ]:
# file_data <- tcltk::tk_choose.files(multi = FALSE)

2. Datos por zona

<font color='tomato'><b> OPCION B:</b> Especificar el nombre de archivo</font>



In [7]:
iFile_02 <- "CU_04_05_17_indicadores_vacunacion.csv"
file_data_02 <- paste0(iPath, iFile_02)

if(file.exists(file_data_02)){
    cat("Se leerán datos del archivo: ", file_data_02)
} else{
    warning("Cuidado: el archivo no existe.")
}


Se leer<U+00E1>n datos del archivo:  Data/Input/CU_04_05_17_indicadores_vacunacion.csv

#### Data file to dataframe

Usar la función adecuada según el formato de entrada (xlsx, csv, json, ...)

In [19]:
data_02 <- read_csv(file_data_02)


Rows: 286 Columns: 26
-- Column specification --------------------------------------------------------
Delimiter: ","
chr  (3): GEOCODIGO, DESBDT, nombre_zona
dbl (23): capacidad_zona, prop_riesgo, tasa_riesgo, tasa_mayores, poblacion_...

i Use `spec()` to retrieve the full column specification for this data.
i Specify the column types or set `show_col_types = FALSE` to quiet this message.


Estructura de  los datos:

In [20]:
data_02 |> glimpse()

Rows: 286
Columns: 26
$ GEOCODIGO         <chr> "001", "002", "003", "004", "005", "006", "007", "00~
$ DESBDT            <chr> "Abrantes", "Acacias", "Adelfas", "Alameda", "Alamed~
$ capacidad_zona    <dbl> 7477, 4482, 7235, 5343, 6816, 5318, 3081, 6843, 3776~
$ prop_riesgo       <dbl> 0.19535423, 0.26312125, 0.24362661, 0.17253854, 0.24~
$ tasa_riesgo       <dbl> 6.867731e-03, 1.091822e-02, 5.821857e-03, 1.797640e-~
$ tasa_mayores      <dbl> 0.033242586, 0.012987683, 0.014193857, 0.009989278, ~
$ poblacion_mayores <dbl> 0.17816190, 0.23909411, 0.22173279, 0.15740288, 0.21~
$ nombre_zona       <chr> "Abrantes", "Acacias", "Adelfas", "Alameda", "Alamed~
$ nsec              <dbl> 23, 11, 19, 18, 18, 15, 10, 17, 11, 12, 14, 17, 8, 2~
$ t3_1              <dbl> 42.31249, 46.15717, 45.30472, 43.70575, 43.31968, 44~
$ t1_1              <dbl> 29872, 17961, 28933, 21393, 27259, 21186, 12367, 274~
$ t2_1              <dbl> 0.5345094, 0.5328775, 0.5383134, 0.4919805, 0.515361~
$ t2_2            

Muestra de datos:

In [12]:
data_02 |> slice_head(n = 5)

GEOCODIGO DESBDT   ano  semana n_vacunas n_citas tmed     prec        
1 001       Abrantes 2021 36     0         0       23.82223 -0.006302348
2 001       Abrantes 2021 37     0         0       20.16001  3.553725804
3 001       Abrantes 2021 38     0         0       18.55106  3.976905218
4 001       Abrantes 2021 39     0         0       18.81539  1.480647232
5 001       Abrantes 2021 40     0         0       17.49447 -0.033790900
  velmedia presMax  benzene   co        no        no2      nox      o3      
1 3.573728 940.9841 0.1713567 0.1680325  4.098371 20.09480 26.48135 50.03434
2 2.494744 940.7610 0.1573829 0.2138607  6.515572 27.42594 37.45944 42.41281
3 3.316148 943.1540 0.1858059 0.2034376  5.477654 20.74836 25.61128 56.29918
4 2.384262 944.6697 0.1486437 0.2399882  9.593391 37.08524 52.43745 46.79483
5 1.850839 944.2973 0.1428037 0.2693450 18.860535 40.19475 74.04903 41.06600
  pm10     pm2.5     so2     
1 17.44765  3.008675 6.861545
2 17.65840 10.083070 6.589638
3 12.84444  7.218588 4.364304
4 16.39590  9.426029 3.123598
5 14.90938  8.131753 1.291137

3. Datos por semana

<font color='tomato'><b> OPCION A:</b> Seleccionar fichero en ventana para mayor comodidad</font> 

Data load using the {tcltk} package. Ucomment the line if not using this option

In [ ]:
# file_data_03 <- tcltk::tk_choose.files(multi = FALSE)

<font color='tomato'><b> OPCION B:</b> Especificar el nombre de archivo</font>



In [15]:
iFile_03 <- "CU_04_05_18_escucha_gripe.csv"
file_data_03 <- paste0(iPath, iFile_03)

if(file.exists(file_data_03)){
    cat("Se leerán datos del archivo: ", file_data_03)
} else{
    warning("Cuidado: el archivo no existe.")
}


Se leer<U+00E1>n datos del archivo:  Data/Input/CU_04_05_18_escucha_gripe.csv

#### Data file to dataframe

Usar la función adecuada según el formato de entrada (xlsx, csv, json, ...)

In [16]:
data_03 <- read_csv(file_data_03)


Rows: 74 Columns: 4
-- Column specification --------------------------------------------------------
Delimiter: ","
dbl (4): ano, semana, tuits_gripe, interes_gripe

i Use `spec()` to retrieve the full column specification for this data.
i Specify the column types or set `show_col_types = FALSE` to quiet this message.


Estructura de  los datos:

In [17]:
data_03 |> glimpse()

Rows: 74
Columns: 4
$ ano           <dbl> 2021, 2021, 2021, 2021, 2021, 2021, 2021, 2021, 2021, 20~
$ semana        <dbl> 36, 37, 38, 39, 40, 41, 42, 43, 44, 45, 46, 47, 48, 49, ~
$ tuits_gripe   <dbl> 97, 79, 112, 143, 112, 130, 254, 190, 198, 160, 206, 280~
$ interes_gripe <dbl> 13, 15, 19, 29, 38, 65, 100, 94, 63, 70, 64, 64, 57, 67,~


Muestra de datos:

In [18]:
data_03 |> slice_head(n = 5)

ano  semana tuits_gripe interes_gripe
1 2021 36      97         13           
2 2021 37      79         15           
3 2021 38     112         19           
4 2021 39     143         29           
5 2021 40     112         38

## ETL Processes

### Import data from: CSV, Excel, Tab, JSON, SQL, and Parquet files

Se han importado en el apartado Data Load anterior:

* Datos por semana y zona (vacunación, meteo y conta)
* Datos por zona (indicadores socieconómicos y sanitarias)
* Datos por semana (búsquedas y tuits)

Incluir apartados si procede para: Extracción de datos (select, filter), Transformación de datos, (mutate, joins, ...). Si es necesario tratar datos perdidos, indicarlo también en NB 09.2

Si no aplica: Estos datos no requieren tareas de este tipo.

#### Data transformation

In [24]:
data <- data_01 |> 
    full_join(data_02, by = c("GEOCODIGO", "DESBDT")) |>
    full_join(data_03, by = c("ano", "semana"))

In [25]:
data |> glimpse()

Rows: 21,736
Columns: 45
$ GEOCODIGO         <chr> "001", "001", "001", "001", "001", "001", "001", "00~
$ DESBDT            <chr> "Abrantes", "Abrantes", "Abrantes", "Abrantes", "Abr~
$ ano               <dbl> 2021, 2021, 2021, 2021, 2021, 2021, 2021, 2021, 2021~
$ semana            <dbl> 36, 37, 38, 39, 40, 41, 42, 43, 44, 45, 46, 47, 48, ~
$ n_vacunas         <dbl> 0, 0, 0, 0, 0, 328, 344, 353, 371, 341, 389, 349, 40~
$ n_citas           <dbl> 0, 0, 0, 0, 0, 305, 327, 328, 341, 334, 373, 332, 37~
$ tmed              <dbl> 23.822231, 20.160014, 18.551058, 18.815387, 17.49447~
$ prec              <dbl> -0.0063023484, 3.5537258042, 3.9769052178, 1.4806472~
$ velmedia          <dbl> 3.573728, 2.494744, 3.316148, 2.384262, 1.850839, 1.~
$ presMax           <dbl> 940.9841, 940.7610, 943.1540, 944.6697, 944.2973, 94~
$ benzene           <dbl> 0.1713567, 0.1573829, 0.1858059, 0.1486437, 0.142803~
$ co                <dbl> 0.1680325, 0.2138607, 0.2034376, 0.2399882, 0.269345~
$ no           

In [27]:
data |> slice_head(n = 5)

GEOCODIGO DESBDT   ano  semana n_vacunas n_citas tmed     prec        
1 001       Abrantes 2021 36     0         0       23.82223 -0.006302348
2 001       Abrantes 2021 37     0         0       20.16001  3.553725804
3 001       Abrantes 2021 38     0         0       18.55106  3.976905218
4 001       Abrantes 2021 39     0         0       18.81539  1.480647232
5 001       Abrantes 2021 40     0         0       17.49447 -0.033790900
  velmedia presMax  ... t7_1       t8_1       t9_1      t10_1     t11_1    
1 3.573728 940.9841 ... 0.04090796 0.02880326 0.2685716 0.1665607 0.4723181
2 2.494744 940.7610 ... 0.04090796 0.02880326 0.2685716 0.1665607 0.4723181
3 3.316148 943.1540 ... 0.04090796 0.02880326 0.2685716 0.1665607 0.4723181
4 2.384262 944.6697 ... 0.04090796 0.02880326 0.2685716 0.1665607 0.4723181
5 1.850839 944.2973 ... 0.04090796 0.02880326 0.2685716 0.1665607 0.4723181
  t12_1     area    densidad_hab_km tuits_gripe interes_gripe
1 0.5637381 1571619 19007.15         97         13           
2 0.5637381 1571619 19007.15         79         15           
3 0.5637381 1571619 19007.15        112         19           
4 0.5637381 1571619 19007.15        143         29           
5 0.5637381 1571619 19007.15        112         38

## Synthetic Data Generation

No aplica

## Fake Data Generation

No aplica

## Open Data

No aplica

## <font color='green'>Data Save</font>

Este proceso, puede copiarse y repetirse en aquellas partes del notebbok que necesiten guardar datos.
Recuerde cambiar las cadenas añadida del fichero para diferenciarlas

<font color='tomato'> Identificamos los datos a guardar</font>

In [28]:
data_to_save <- data


<font color='tomato'>Estructura de nombre de archivos:</font>

* Código del caso de uso, por ejemplo "CU_04"
* Número del proceso que lo genera, por ejemplo "_05".
* Número de la tarea que lo genera, por ejemplo "_01"
* En caso de generarse varios ficheros en la misma tarea, llevarán _01 _02 ... después
* Nombre: identificativo de "properData", por ejemplo "_zonasgeo"
* Extensión del archivo

Ejemplo: "CU_04_05_01_01_zonasgeo.json, primer fichero que se genera en la tarea 01 del proceso 05 (Data Collection) para el caso de uso 04 (vacunas)

Importante mantener los guiones bajos antes de proceso, tarea, archivo y nombre

### Proceso 05

In [29]:
caso <- "CU_04"
proceso <- '_05'
tarea <- "_19"
archivo <- ""
proper <- "_vacunacion_gripe_completo"
extension <- ".csv"

<font color='tomato'><b> OPCION A:</b> Uso del paquete "tcltk" para mayor comodidad</font>

* Buscar carpeta, escribir nombre de archivo SIN extensión (se especifica en el código)
* Especificar sufijo2 si es necesario
* Cambiar datos por datos_xx si es necesario

In [ ]:
# file_save <- paste0(caso, proceso, tarea, tcltk::tkgetSaveFile(), proper, extension) 
# path_out <- paste0(oPath, file_save)
# write_csv(data_to_save, path_out)

# cat('File saved as: ')
# path_out

<font color='tomato'><b> OPCION B:</b> Especificar el nombre de archivo</font>

- Los ficheros de salida del proceso van siempre a Data/Output/.  

In [30]:
file_save <- paste0(caso, proceso, tarea, archivo, proper, extension) 
path_out <- paste0(oPath, file_save)
write_csv(data_to_save, path_out)

cat('File saved as: ')
path_out

File saved as: 

[1] "Data/Output/CU_04_05_19_vacunacion_gripe_completo.csv"

#### Copia del fichero a Input

Si el archivo se va a usar en otros notebooks, copiar a la carpeta Input

In [31]:
path_in <- paste0(iPath, file_save)
file.copy(path_out, path_in, overwrite = TRUE)

[1] TRUE

## <font color=' #2874a6 '>Main Conclusions</font>

<font color='steelblue'>List and describe the general conclusions of the analysis carried out​.</font>

### Prerequisites

Para que funcione este código se necesita:

* Las rutas de archivos `Data/Input` y `Data/Output` deben existir (relativas a la ruta del _notebook_)
* El paquete tcltk instalado para seleccionar archivos interactivamente. No se necesita en producción.
* Los paquetes readr, dplyr deben estar instalados.


### Configuration Management

This notebook has been tested with the following versions of R and packages. It cannot be assured that later versions work in the same way:
* R 4.2.2
* tcltk 4.2.2
* readr 2.1.3
* dplyr 1.0.10

### Data structures

#### Objeto `data`

- Hay 21736 filas con informaci<U+00F3>n de las siguientes variables:
	GEOCODIGO, DESBDT, ano, semana, n_vacunas, n_citas, tmed, prec, velmedia, presMax, benzene, co, no, no2, nox, o3, pm10, pm2.5, so2, capacidad_zona, prop_riesgo, tasa_riesgo, tasa_mayores, poblacion_mayores, nombre_zona, nsec, t3_1, t1_1, t2_1, t2_2, t4_1, t4_2, t4_3, t5_1, t6_1, t7_1, t8_1, t9_1, t10_1, t11_1, t12_1, area, densidad_hab_km, tuits_gripe, interes_gripe


#### Observaciones generales sobre los datos

- Los datos se repiten en aquellas columnas que no tienen la misma resolución que el conjunto más amplio, es decir:
    - Los datos de escucha son los mismos para todas las zonas la misma semana
    - Los datos de indicadores son los mismos para todas las semanas en cada zona
- El fichero incluye algunos datos reales y otros sintéticos
- Algunas tareas del resto de procesos de la dimensión Data Processing se han realizado en este proceso de Data Collection por motivos de eficiencia. En los procesos correspondientes se reportarán estas acciones realizadas

### Consideraciones para despliegue en piloto

* Este es el fichero principal que contiene todos los datos del caso
* Si llegaran datos reales solicitados a la empresa, habría que sustituir los sintéticos por los reales
* Para la representación en mapas, se deben combinar los datos del csv con los datos del json en la implementación que se haga

### Consideraciones para despliegue en producción

* Se deben crear los procesos ETL en producción necesarios para que los datos de entrada estén actualizados
* Se deben conseguir datos reales para sustituir los sintéticos
* Para la representación en mapas, se deben combinar los datos del csv con los datos del json de la primera trea en la implementación que se haga

## <font color=' #2874a6 '>Main Actions</font>

#### <font color=' #2874a6 '>Acciones done</font>

<font color='steelblue'>Indicate the actions that have been carried out in this process</font>

- Se han unido las tablas de datos del caso de uso en un único archivo

#### <font color=' #2874a6 '>Acctions to perform</font>

<font color='steelblue'>Indicate the actions that must be carried out in subsequent processes</font>

- Se deben realizar el resto de procesos de la dimensión Data Processing

## <font color='RED'>CODE TO DEPLOY (PILOT)</font>

A continuación se incluirá el código que deba ser llevado a despliegue para producción, dado que se entiende efectúa operaciones necesarias sobre los datos en la ejecución del prototipo

<font color=' red '><b>Description</b></font>

- No hay código que desplegar en el piloto, ya que estos datos son estáticos o en todo caso cambian con muy poca frecuencia, altamente improbable durante el proyecto.
- El fichero deberá estar incluido en la ruta de datos de la implementación


<font color=' red '><b>CODE</b></font>

In [ ]:
# incluir código